src/main/java/fi/aalto/cs/apluscourses/model/CoursePluginVersion.java

```diff
@@ -0,0 +1,51 @@
+package fi.aalto.cs.apluscourses.model;
+
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONObject;
+
+public class CoursePluginVersion {
+
+  public enum Status {
+    VALID,
+    UPDATE_OPTIONAL,
+    UPDATE_REQUIRED
+  }
+
+  // the major and minor version of the current plugin build
+  private static final int MAJOR_VERSION = 1;
+  private static final int MINOR_VERSION = 0;
+  public static final CoursePluginVersion CURRENT_VERSION =
+      new CoursePluginVersion(MAJOR_VERSION, MINOR_VERSION);
+
+  private final int major;
+  private final int minor;
+
+  /**
+   * Constructs a course plugin version directly from arguments.
+   */
+  public CoursePluginVersion(int major, int minor) {
+    this.major = major;
+    this.minor = minor;
+  }
+
+  /**
+   * Constructs a course plugin version from a JSON object.
+   */
+  public CoursePluginVersion(@NotNull JSONObject jsonObject) {
+    this(jsonObject.getInt("major"),
+         jsonObject.getInt("minor"));
+  }
+
+  /**
+   * Determines whether the current plugin version is recent enough for the given plugin version.
+   */
+  public Status checkVersion() {
+    if (MAJOR_VERSION < major) {
+      return Status.UPDATE_REQUIRED;
+    }
+    if (MAJOR_VERSION > major) {
+      return Status.VALID;
+    }
+    return MINOR_VERSION >= minor ? Status.VALID : Status.UPDATE_OPTIONAL;
+  }
```

Comment:
Correct me if I'm wrong: this seems to assume that newer plugin versions are compatible with older courses, even if there has been a change in the major version number? I think it would be safer to enable breaking changes both ways, i.e. the plugin is considered compatible with the course only if the major versions match.

Unrelated: it might be nice to have this method take another version parameter instead of using the static one, something like `Status isCompatibleWith(CoursePluginVersion other)`, then it would be more generic and unit testable

Reply:
Sorry, the name `isCompatibleWith` is of course a bad idea, because it doesn't return a boolean. But taking the compared-to version as a parameter would still be a good idea

Reply:
Correct, new plugin versions are compatible with all previous minor releases of the same major AND all previous major releases, period.

What should the user do if the course requires an older major version? Do they even have an option for downgrading the A+ plugin? I haven't really thought about it but it sounds like a flaw on our side if we decide to drop support for old but currently running courses. (The finished courses will eventually be removed from the course selection anyway, like O1 2020 will stop showing up once 2021 is out).

Reply:
Downgrading the plugin is possible, but it's not trivial: uninstall the plugin, download a ZIP of the correct version from the JetBrains plugin website, and install it using "install plugin from disk".

> I haven't really thought about it but it sounds like a flaw on our side if we decide to drop support for old but currently running courses

If the old course is still supported, then why would it have a different major version?

The way I see it, with the current approach we either support old courses until eternity, or the old courses break in some way but the user doesn't get a coherent incompatibility error notification.

I also haven't carefully considered this through so I might be missing something

Reply:
> The way I see it, with the current approach we either support old courses until eternity

Well... that's what I thought ;v or at least not until eternity, but as long as the course is present in the course selection list. To be honest, I was actually under the impression that old courses disappear from A+ entirely - maybe I was biased by this.

If someone manually enters the O1 2020 link (even though the course is gone from the course selection list), they're already assumed to be an "advanced user" - normal people don't even know what a course URL is!

But let's ignore what I wrote above and consider this: when we release an update with the IDE activities, we'll surely bump the major version from 1 to 2 (version 1 won't be able to run the IDE activities). But this does not mean that a version 2 plugin cannot open the O1 2020 course (which runs on version 1, without IDE activities). So in reality, the course would need to feature a "minimum compatible version" and a "maximum compatible version", with the latter manually updated when we release a truly breaking change (like a change to the course JSON structure). That's unreasonable both for us and the course administrators.

Reply:
For future reference: after a duel in the daily, we decided to proceed with Pawel's plan.

Reply:
> when we release an update with the IDE activities, we'll surely bump the major version from 1 to 2 (version 1 won't be able to run the IDE activities)

I'm confused.

Are we talking about something else than the plugin's version (that is currently 2.0)?

I see the schema of course config file as an _interface_ that the plug-in _realizes_ and courses _use_. Thus, it makes sense to refer to the _version of the plugin_ in the course config file to solve the maintainability question.

Adding support to the assessable IDE activities isn't a breaking change: the plugin is still compatible with the older courses. Thus, the major version of the plugin shouldn't be increased because of that. Let's assume it's gonna be v2.2 that first introduces IDE-acts. Then a course that requires them refers to v2.2 in its course config file. Plugin versions 2.2, 2.3, 2.4, ... should all consider themselves compatible with that course. Then again, v2.1 and v3.0 should not, the former because it doesn't have the required functionality (IDE-acts, in this case), the latter because it contains some changes that breaks the contract the author of that course relied on.

Reply:
> Are we talking about something else than the plugin's version (that is currently 2.0)?

Yes - we're talking about the version specified in `CoursePluginVersion` class as MAJOR_VERSION and MINOR_VERSION constants. This is completely unrelated to the plugin build version (currently "v2.0").

> Adding support to the assessable IDE activities isn't a breaking change

Indeed, it does not break backwards compatibility. However the CoursePluginVersion should have its major version increased, because that's how the presence of the new required feature (IDE activities) will be distinguished. Otherwise the new O1 course (with IDE activities) won't have any way of actually requiring a new plugin version.

That's how this would work in practice:
1. Let's say we release v2.3, v2.4, v2.5 without IDE activities. These have MAJOR_VERSION = 1.
2. In v2.6 IDE activities are fully working and released. MAJOR_VERSION = 2.
3. The O1 2021 course now specifies that major version 2 is required. This means that all users running v2.6 get accepted, and all users running v2.5 or older get prompted to update.
4. If we ever release some new important feature for other courses, we'll bump MAJOR_VERSION to 3. O1 2021 will still work though because its required major version is lower than 3.

So the `CoursePluginVersion` is more like a "feature set version" rather than a "backwards compatibility version".

If we ever release an update that destroyes backwards compatibility with existing courses (courses with lower required major version no longer work) then **this ticket doesn't solve this problem**. Running courses should not be modified like that anyway, and we've (I and Nikolas) somewhat agreed on the fact that after a course is removed from the plugin's course selection list, it will most likely be never instantiated anymore. This would also mark the end of support for these courses.

Reply:
I appreciate your investment in the issue! I see you have thought this trough very well and have a strong understanding on the issue. By keeping challenging your ideas, I don't want to shoot them down, but make sure we are talking about the same things and are solving the same problems. I hope you have energy to answer my questions below.

> we're talking about the version specified in CoursePluginVersion class as MAJOR_VERSION and MINOR_VERSION constants. This is completely unrelated to the plugin build version (currently "v2.0").

Does this then leave "the plugin build version" out of any semantic meaning? Initially, it was our idea that the plugin version itself is what is used to resolve the compatibility with courses. Now, this sounds to me that we are introducing another version (version of the course-config language/fileformat?). Is this what you intended?

> the CoursePluginVersion should have its major version increased, because that's how the presence of the new required feature (IDE activities) will be distinguished. Otherwise the new O1 course (with IDE activities) won't have any way of actually requiring a new plugin version.

Why wouldn't we use minor version for additions that aren't breaking? Courses would then require plugin version _A.B_ with some major version _A_ and minor version _B_. Plugin versions _a.b_ where _a=A_ and _b>=B_ would then be compatible. Usually, that's what major/minor versions are for (such as in [SemVer standard](https://semver.org/)).

>That's how this would work in practice:
>
> 1. Let's say we release v2.3, v2.4, v2.5 without IDE activities. These have MAJOR_VERSION = 1.
> 2. In v2.6 IDE activities are fully working and released. MAJOR_VERSION = 2.
> 3. The O1 2021 course now specifies that major version 2 is required. This means that all users running v2.6 get accepted, and all users running v2.5 or older get prompted to update.
> 4. If we ever release some new important feature for other courses, we'll bump MAJOR_VERSION to 3. O1 2021 will still work though because its required major version is lower than 3.

Does MINOR_VERSION carry any meaning here?

> So the CoursePluginVersion is more like a "feature set version" rather than a "backwards compatibility version".

Could it be both if minor version is the "feature set version" and major version is the "backwards compatibility version"?

> This would also mark the end of support for these courses, when changes that break backwards compatibility can be released.

Course config files can keep circling around even if their courses have ended. A teacher can start a new instance of an old course by copying the old course config file. I'd see it ideal that they could just try loading the old file with the newest plugin version and, by seeing no warnings, make sure it is still supported by the plugin (meaning there's no breaking changes in-between). If they want to include new features they can just upgrade the required version to the newest one (or, if they are super pedantic, they can find the actual version that first introduced that feature and use that). What do you think?

Reply:
> Does this then leave "the plugin build version" out of any semantic meaning?

To be honest, it seems so. Perhaps I really haven't thought this thoroughly enough, but for example when v2.0 was released, there really weren't any breaking changes introduced despite major version increasing. Right now v1.8 and v2.0 are simultenously in active use in the same course. That's probably what influenced my decision of introducing a parallel version class.

> Why wouldn't we use minor version for additions that aren't breaking? Could it be both if minor version is the "feature set version" and major version is the "backwards compatibility version"?

Yes, this is possible. The thing is that when there's a mismatch between minor versions, a warning is shown, not an error. So the user might proceed anyway using a version that does not have IDE activities and be confused (perhaps exceptions due to unknown assignment types/unknown JSON fields or perhaps the assignments simply won't show up - either way, it's unspecified behaviour - something that we're trying to say goodbye to with this ticket).

Perhaps minor version should actually cause an error? But then it's no longer a truly "minor" version... the more I think about it, the more confused I become ;<

Reply:
No wonder we are confused here -- these are by no means easy things. 🤷‍♂️

>when v2.0 was released, there really weren't any breaking changes introduced despite major version increasing. Right now v1.8 and v2.0 are simultenously in active use in the same course. That's probably what influenced my decision of introducing a parallel version class.

You're absolutely correct here. We jumped to v2.0, I'd say, more for "marketing" reasons than technical ones. I'll add my review to this PR later today and try to consider from the perspective I got from your answers. So I can give feedback of the code. Let's consider then later (in a meeting?) which approach we want to take with the versioning. 👍

Reply:
I think 2.0 does count as breaking, because it introduces a *mandatory* `languages` property to the course configuration